python/ray/tune/trial_runner.py

```diff
@@ -527,6 +535,39 @@ def _process_trial(self, trial):
                 result = trial.last_result
                 result.update(done=True)
 
+            # Check if any of the required metrics was not reported
+            # in the last result. If the only item is `done=True`, this
+            # means that no result was ever received and the trial just
+            # returned. This is also okay and will not raise an error.
+            if int(os.environ.get("TUNE_DISABLE_STRICT_METRIC_CHECKING",
+                                  0)) != 1 and (len(result) > 1
+                                                or "done" not in result):
+                base_metric = self._metric
+                scheduler_metric = self._scheduler_alg.metric
+                search_metric = self._search_alg.metric
+
+                if base_metric and base_metric not in result:
+                    report_metric = base_metric
+                    location = "tune.run()"
+                elif scheduler_metric and scheduler_metric not in result:
+                    report_metric = scheduler_metric
+                    location = type(self._scheduler_alg).__name__
+                elif search_metric and search_metric not in result:
+                    report_metric = search_metric
+                    location = type(self._search_alg).__name__
+                else:
+                    report_metric = None
+                    location = None
+
+                if report_metric:
+                    raise ValueError(
+                        "Trial returned a result which did not include the "
+                        "specified metric `{}` that `{}` expects. "
+                        "Make sure your calls to `tune.report()` include the "
+                        "metric, or set the "
+                        "TUNE_DISABLE_STRICT_METRIC_CHECKING "
```

Comment:
any reason not to make this a tune.run variable?

Reply:
I think it is a sensible default to have strict metric checking enabled and it should only be disabled in very specific (and rare) cases. This way we can keep the tune.run API clean.

Reply:
like `strict_metric_check=False`

Reply:
So IMO the question is how often users want to report metrics that are not acted upon to Tune and why they would do that. It could be for (e.g. tensorboard) logging and it could be if the search algorithm and scheduler should act on different metrics that are calculated in different steps of the training function (e.g. one during training, one during evaluation). My impression is that these cases are rather rare, but I might be mistaken.

Reply:
For the second case, we could actually check if at least one of the expected metrics has been reported. That would be a not-so-strict metric enforcement.

Reply:
OK, then is it possible to make sure we don't set this env var in any of the examples? (i.e., I think pbt_transformers is one of the failing examples, along with xgboost_example which currently has the env var set)

The examples should be representative of very commonplace usage (or at least, a baseline), and setting an env var for those would be undesirable.

Reply:
(sorry my github isn't actually refreshing fast enough so I'm missing your messages heh)

Reply:
RE: reporting non-actionable metrics: logging is a good use case (see rllib); choosing among different metrics to optimize at runtime is also a good use-case (i.e., xgboost). 

Basically, having strict check be the default is fine, but we should be sure that all of our examples represent common case rather than edge case.

Reply:
Yeah, I fixed the transformer example and can update the xgboost one right away